hcxpioff.1
.TH HCXPIOF "1" "Oct 2021" "HCXPIOF 6.2.4" "Small tool to capture packets from wlan devices"
.\" Text automatically generated by txt2man
.RS
.SH NAME
\fBhcxpiof \fP- Turns Raspberry Pi off via GPIO switch
\fB
.SH SYNOPSIS
.nf
.fam C
\fBhcxpiof\fP [\fIOPTIONS\fP]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
Turns Raspberry Pi off via GPIO switch
.SH OPTIONS
.TP
.B
\fB-h\fP
show this help
.TP
.B
\fB-v\fP
show version
.PP
\fB--gpio_button\fP=<digit>
Raspberry Pi GPIO pin number of button (2\.\.\.27)
default = GPIO not in use
.PP
\fB--gpio_statusled\fP=<digit>
Raspberry Pi GPIO number of status LED (2\.\.\.27)
default = GPIO not in use
.PP
\fB--tot\fP=<digit>
enable timeout timer in minutes (minimum = 2 minutes)
hcxpioff will terminate if tot reached
.PP
\fB--reboot\fP once hcxpioff terminated, reboot system
default: power off system
.PP
\fB--help\fP display help and exit
.PP
\fB--version\fP show version
.SH NOTES
run gpio readall to print a table of all accessible pins and their numbers
(wiringPi, BCM_GPIO and physical pin numbers)
.PP
press the button to power off
hardware modification is necessary, read more:
https://github.com/ZerBea/hcxdumptool/tree/master/docs
.SH AUTHOR
Written by ZeroBeat <zerobeat@gmx.de>
.RE
.PP
This manual page was written by Paulo Roberto Alves de Oliveira (aka kretcheu)
<kretcheu@gmail.com> for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright © 2000-2021 ZeroBeat <zerobeat@gmx.de>
License MIT


changelog
24.10.2021
==========
hcxdumptool: added new option to ignore ioctl() warnings
--force_interface                  : ignore all ioctl() warnings
                                     do not report issues, if attacks or channel switch is not working as expected


23.10.2021
==========
added generic hcxdumptool.1 man page 


22.10.2021
==========
moved install dir from usr/local/bin to /usr/bin


14.09.2021
==========
release v6.2.4
This is the last version, supporting OpenSSL 1.1
Next version 6.3.0 will need OpenSSL 3.0.0


01.09.2021
==========
hcxdumptool: added new option to the next EAP request entry contained in the sequence passed to
             --eapreq with a type matching the Desired Auth Type requested in a client's Legacy NAK response
--eapreq_follownak                 : jump to Auth Type requested by client in Legacy Nak response, if type available in remaining request sequence


29.08.2021
==========
hcxdumptool: GPIO flash intervall not longer hardcoded
--gpio_statusled_intervall=<digit> : Raspberry Pi GPIO LED flash intervall
                                     default = flash every 5 seconds


15.08.2021
==========
hcxdumptool: allow MAC instead of filter list
--filterlist_ap=<file or MAC>      : ACCESS POINT MAC or MAC filter list
                                     format: 112233445566, 11:22:33:44:55:66, 11-22-33-44-55-66 # comment
                                     maximum entries 256
                                     run first --do_rcascan to retrieve information about the target
--filterlist_client=<file or MAC>  : CLIENT MAC or MAC filter list
                                     format: 112233445566, 11:22:33:44:55:66, 11-22-33-44-55-66 # comment
                                     maximum entries 256
                                     due to MAC randomization of the CLIENT, it does not always work!


17.05.2021
==========
release v 6.1.6
get ready for gcc 11.1.0


18.03.2021
==========
hcxdumptool: added VENDOR filter
--filterlist_ap_vendor=<file>      : ACCESS POINT MAC filter list by VENDOR
                                     format: 112233, 11:22:33, 11-22-33 # comment
                                     maximum entries 256
                                     run first --do_rcascan to retrieve information about the target
--filterlist_client_VENDOR=<file>  : CLIENT MAC filter list
                                     format: 112233, 11:22:33, 11-22-33 # comment
                                     maximum entries 256
                                     due to MAC randomization of the CLIENT, it does not always work!


09.03.2021
==========
release v 6.1.6
get ready for OpenSSL 3.0


05.03.2021
==========
hcxdumptool: removed preamble option, because it is ignored by most wlan adapters


04.03.2021
==========
hcxdumptool: added new option to use short preamble
--short_preamble                   : use short preamble
                                     default: use long preamble
that may not work on older devices!


23.02.2021
==========
hcxdumptool: allow output to stdout
-o <dump file> : output file in pcapng format, filename '-' outputs to stdout
                 including radiotap header (LINKTYPE_IEEE802_11_RADIOTAP)


14.02.2021
==========
sll_pkttype = PACKET_OUTGOING is completey ingnored in promiscuous mode
At least I found a way to detect outgoing packets.
Luckily IEEE80211_RADIOTAP_TX_FLAGS is set on outgoing packets. If we check this, we are able to ignore them.
That save us CPU time.


13.02.2021
==========
hcxdumptool: added WATCHDOG - we need that to determine if USB device died
             Raspberry Pi LED will flash twice if now packet received turing the last past 600 seconds
             print AGE of last received packet in status


09.02.2021
==========
started to prepare to use openssl 3.0
Many structures have been made opaque in OpenSSL 3.0 since OpenSSL 1.0.2
https://wiki.openssl.org/index.php/OpenSSL_3.0
it is recommended to upgrade to at least to OpenSSL 1.1 


06.02.2021
==========
hcxdumptool: fixed handling of BE (according to hcxpcapngtool fixes)


18.01.2021
==========
hcxdumptool: added auto channel/auto band.
             therefore default channel list and predefined channel lists changed!

-c <digit>     : set channel (1,2,3, ...)
                 default: auto channel/auto band
                 maximum entries: 255
                 standard 802.11 channels (depends on device, driver and world regulatory domain):
                 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14
                 36, 38, 40, 44, 48, 52, 56, 60, 64
                 100, 104, 108, 112, 116, 120, 124, 128
                 132, 136, 140, 144, 149, 153, 157, 161, 165
                 201, 205, 209, 213, 217, 221, 225, 229, 233
-s <digit>     : set predefined scanlist
                 0 = auto channel/auto band (default)
                 1 = 1,6,11,3,5,1,6,11,2,4,1,6,11,7,9,1,6,11,8,10,1,6,11,12,13 (optimized 2.4GHz)
                 2 = 1,2,3,4,5,6,7,8,9,10,11,12,13 (standard 2.4 GHz)
                 3 = 36,40,44,48,52,56,60,64,100,104,108,112,116,120,124,128,132,136,140,149,153,157,161,165 (standard 5GHz)
                 4 = 1,2,3,4,5,6,7,8,9,10,11,12,13,36,40,44,48,52,56,60,64,100,104,108,112,116,120,124,128,132,136,140,149,153,157,161,165 (standard 2.4GHz/5GHz)
                 5 = 201,205,209,213,217,221,225,229,233 (standard 6GHz)


16.01.2021
==========
hcxdumptool: added new option - rcascan ranking is no longer hard coded
--rcascan_order=digit>             : rcascan sorting order:
                                      0 = sort by HIT (PROBERESPONSE) (default)
                                      1 = sort by COUNT (BEACON)
                                      2 = sort by CHANNEL


14.01.2021
==========
release v 6.1.5


03.01.2021
==========
hcxdummptool: added new option to limit status of rcascan to n lines
--rcascan_max=digit>               : show only n highest ranking lines

hcxpioff.c
#define _GNU_SOURCE
#include <getopt.h>
#include <signal.h>
#include <fcntl.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>

#include "include/rpigpio.h"

#define HCX_GPIO_BUTTON		1
#define HCX_GPIO_STATUSLED	2
#define HCX_TOT			3
#define HCX_REBOOT		4
#define HCX_POWER_OFF		5
#define HCX_HELP		'h'
#define HCX_VERSION		'v'
/*===========================================================================*/
/* global var */

static int gpiostatusled;
static int gpiobutton;
static struct timespec sleepled;
struct timeval tv;
struct timeval tvtot;

static bool poweroffflag;
static bool rebootflag;
/*===========================================================================*/
__attribute__ ((noreturn))
static void globalclose()
{
if(gpiostatusled > 0) GPIO_SET = 1 << gpiostatusled;
if(poweroffflag == true)
	{
	if(system("poweroff") != 0)
		{
		printf("can't power off\n");
		exit(EXIT_FAILURE);
		}
	}
if(rebootflag == true)
	{
	if(system("reboot") != 0)
		{
		printf("can't reboot\n");
		exit(EXIT_FAILURE);
		}
	}
exit(EXIT_SUCCESS);
}
/*===========================================================================*/
static inline size_t chop(char *buffer, size_t len)
{
static char *ptr;

ptr = buffer +len -1;
while(len)
	{
	if (*ptr != '\n') break;
	*ptr-- = 0;
	len--;
	}
while(len)
	{
	if (*ptr != '\r') break;
	*ptr-- = 0;
	len--;
	}
return len;
}
/*---------------------------------------------------------------------------*/
static inline int fgetline(FILE *inputstream, size_t size, char *buffer)
{
static size_t len;
static char *buffptr;

if(feof(inputstream)) return -1;
buffptr = fgets (buffer, size, inputstream);
if(buffptr == NULL) return -1;
len = strlen(buffptr);
len = chop(buffptr, len);
return len;
}
/*===========================================================================*/
static bool initgpio(int gpioperi)
{
static int fd_mem;

fd_mem = open("/dev/mem", O_RDWR|O_SYNC);
if(fd_mem < 0)
	{
	fprintf(stderr, "failed to get device memory\n");
	return false;
	}
gpio_map = mmap(NULL, BLOCK_SIZE, PROT_READ|PROT_WRITE, MAP_SHARED, fd_mem, GPIO_BASE +gpioperi);
close(fd_mem);
if(gpio_map == MAP_FAILED)
	{
	fprintf(stderr, "failed to map GPIO memory\n");
	return false;
	}
gpio = (volatile unsigned *)gpio_map;
return true;
}
/*===========================================================================*/
static int getrpirev()
{
static FILE *fh_rpi;
static int len;
static int rpi = 0;
static int rev = 0;
static int gpioperibase = 0;
static char *revptr = NULL;
static const char *revstr = "Revision";
static const char *hwstr = "Hardware";
static const char *snstr = "Serial";
static char linein[128];

fh_rpi = fopen("/proc/cpuinfo", "r");
if(fh_rpi == NULL)
	{
	perror("failed to retrieve cpuinfo");
	return gpioperibase;
	}
while(1)
	{
	if((len = fgetline(fh_rpi, 128, linein)) == -1) break;
	if(len < 15) continue;
	if(memcmp(&linein, hwstr, 8) == 0)
		{
		rpi |= 1;
		continue;
		}
	if(memcmp(&linein, revstr, 8) == 0)
		{
		rpirevision = strtol(&linein[len -6], &revptr, 16);
		if((revptr - linein) == len)
			{
			rev = (rpirevision >> 4) &0xff;
			if(rev <= 3)
				{
				gpioperibase = GPIO_PERI_BASE_OLD;
				rpi |= 2;
				continue;
				}
			if(rev == 0x09)
				{
				gpioperibase = GPIO_PERI_BASE_OLD;
				rpi |= 2;
				continue;
				}
			if(rev == 0x0c)
				{
				gpioperibase = GPIO_PERI_BASE_OLD;
				rpi |= 2;
				continue;
				}
			if((rev == 0x04) || (rev == 0x08) || (rev == 0x0d) || (rev == 0x00e) || (rev == 0x011))
				{
				gpioperibase = GPIO_PERI_BASE_NEW;
				rpi |= 2;
				continue;
				}
			continue;
			}
		rpirevision = strtol(&linein[len -4], &revptr, 16);
		if((revptr - linein) == len)
			{
			if((rpirevision < 0x02) || (rpirevision > 0x15)) continue;
			if((rpirevision == 0x11) || (rpirevision == 0x14)) continue;
			gpioperibase = GPIO_PERI_BASE_OLD;
			rpi |= 2;
			}
		continue;
		}
	if(memcmp(&linein, snstr, 6) == 0)
		{
		rpi |= 4;
		continue;
		}
	}
fclose(fh_rpi);
if(rpi < 0x7) return 0;
return gpioperibase;
}
/*===========================================================================*/
static bool globalinit()
{
static int gpiobasemem = 0;

sleepled.tv_sec = 0;
sleepled.tv_nsec = GPIO_LED_DELAY;
if((gpiobutton > 0) || (gpiostatusled > 0))
	{
	if(gpiobutton == gpiostatusled)
		{
		fprintf(stderr, "same value for wpi_button and wpi_statusled is not allowed\n");
		return false;
		}
	gpiobasemem = getrpirev();
	if(gpiobasemem == 0)
		{
		fprintf(stderr, "failed to locate GPIO\n");
		return false;
		}
	if(initgpio(gpiobasemem) == false)
		{
		fprintf(stderr, "failed to init GPIO\n");
		return false;
		}
	if(gpiostatusled > 0)
		{
		INP_GPIO(gpiostatusled);
		OUT_GPIO(gpiostatusled);
		}
	if(gpiobutton > 0) INP_GPIO(gpiobutton);
	}
return true;
}
/*===========================================================================*/
static void ledflash()
{
if(gpiostatusled == 0) return;
GPIO_SET = 1 << gpiostatusled;
nanosleep(&sleepled, NULL);
GPIO_CLR = 1 << gpiostatusled;
nanosleep(&sleepled, NULL);
GPIO_SET = 1 << gpiostatusled;
nanosleep(&sleepled, NULL);
GPIO_CLR = 1 << gpiostatusled;
nanosleep(&sleepled, NULL);
return;
}
/*===========================================================================*/
__attribute__ ((noreturn))
static void waitloop()
{
static int count = 1;
while(1)
	{
	if(GET_GPIO(gpiobutton) > 0) globalclose();
	gettimeofday(&tv, NULL);
	if(tv.tv_sec >= tvtot.tv_sec) globalclose();
	if((count %5) == 0) ledflash();
	count++;
	sleep(1);
	}
}
/*===========================================================================*/
__attribute__ ((noreturn))
static inline void version(char *eigenname)
{
printf("%s %s (wpi version) (C) %s ZeroBeat\n", eigenname, VERSION_TAG, VERSION_YEAR);
exit(EXIT_SUCCESS);
}
/*---------------------------------------------------------------------------*/
__attribute__ ((noreturn))
static inline void usage(char *eigenname)
{
printf("%s %s (wpi version) (C) %s ZeroBeat\n"
	"usage  : %s <options>\n"
	"         press the button to power off\n"
	"         hardware modification is necessary, read more:\n"
	"         https://github.com/ZerBea/hcxdumptool/tree/master/docs\n" 
	"\n"
	"options:\n"
	"-h       : show this help\n"
	"-v       : show version\n"
	"\n"
	"--gpio_button=<digit>    : Raspberry Pi GPIO pin number of button (2...27)\n"
	"                           default = GPIO not in use\n"
	"--gpio_statusled=<digit> : Raspberry Pi GPIO number of status LED (2...27)\n"
	"                           default = GPIO not in use\n"
	"--tot=<digit>            : enable timeout timer in minutes (minimum = 2 minutes)\n"
	"                         : hcxpioff will terminate if tot reached\n"
	"--reboot                 : once hcxpioff terminated, reboot system\n"
	"                         : default: power off system\n"
	"--help                   : show this help\n"
	"--version                : show version\n"
	"\n"
	"run gpio readall to print a table of all accessible pins and their numbers\n"
	"(wiringPi, BCM_GPIO and physical pin numbers)\n"
	"\n",
	eigenname, VERSION_TAG, VERSION_YEAR, eigenname);
exit(EXIT_SUCCESS);
}
/*---------------------------------------------------------------------------*/
__attribute__ ((noreturn))
static inline void usageerror(char *eigenname)
{
printf("%s %s (wpi version) (C) %s by ZeroBeat\n"
	"usage: %s -h for help\n", eigenname, VERSION_TAG, VERSION_YEAR, eigenname);
exit(EXIT_FAILURE);
}
/*===========================================================================*/
int main(int argc, char *argv[])
{
static int auswahl;
static int index;
static long int totvalue;

static const char *short_options = "hv";
static const struct option long_options[] =
{
	{"gpio_button",		required_argument,	NULL,	HCX_GPIO_BUTTON},
	{"gpio_statusled",	required_argument,	NULL,	HCX_GPIO_STATUSLED},
	{"tot",			required_argument,	NULL,	HCX_TOT},
	{"reboot",		no_argument,		NULL,	HCX_REBOOT},
	{"version",		no_argument,		NULL,	HCX_VERSION},
	{"help",		no_argument,		NULL,	HCX_HELP},
	{NULL,			0,			NULL,	0}
};

auswahl = -1;
index = 0;
optind = 1;
optopt = 0;
gpiostatusled = 0;
gpiobutton = 0;
rebootflag = false;
poweroffflag = true;
tvtot.tv_sec = 2147483647L;
tvtot.tv_usec = 0;

while((auswahl = getopt_long(argc, argv, short_options, long_options, &index)) != -1)
	{
	switch (auswahl)
		{
		case HCX_GPIO_BUTTON:
		gpiobutton = strtoll(optarg, NULL, 10);
		if((gpiobutton < 2) || (gpiobutton > 27))
			{
			fprintf(stderr, "only 2...27 allowed\n");
			exit(EXIT_FAILURE);
			}
		break;

		case HCX_GPIO_STATUSLED:
		gpiostatusled = strtoll(optarg, NULL, 10);
		if((gpiostatusled < 2) || (gpiostatusled > 27))
			{
			fprintf(stderr, "only 2...27 allowed\n");
			exit(EXIT_FAILURE);
			}
		break;

		case HCX_TOT:
		totvalue = strtol(optarg, NULL, 10);
		if(totvalue < 2)
			{
			fprintf(stderr, "tot must be >= 2 (minutes)\n");
			exit(EXIT_FAILURE);
			}
		gettimeofday(&tvtot, NULL);
		tvtot.tv_sec += totvalue *60;
		break;

		case HCX_REBOOT:
		rebootflag = true;
		poweroffflag = false;
		break;

		case HCX_HELP:
		usage(basename(argv[0]));
		break;

		case HCX_VERSION:
		version(basename(argv[0]));
		break;

		case '?':
		usageerror(basename(argv[0]));
		break;
		}
	}

if(globalinit() == false)
	{
	exit(EXIT_FAILURE);
	}

waitloop();

return EXIT_SUCCESS;
}
/*===========================================================================*/
